### PR TITLE
Auto shout out and auto go-live posts

### LINKED ISSUE #25: Auto shout out and go-live
# Feature
Automatically shout out streamers across all servers that has auto shout out enabled. This is a controlled way of promoting other streamers in your server!
Additionally, automatically post go-live posts in servers where it's enabled.

## Acceptance Criteria
- [x] Sushi Bot shall allow server administrators to configure auto shout outs.
- [x] Sushi Bot shall shout out users who go live.
  - [x] This will only be done in servers that has the feature configured.
  - [x] The shout out message may be set by the server administrators.
- [x] Sushi Bot shall allow server administrators to configure auto go-live posts.
- [x] Sushi Bot shall post a go-live message whenever the server owner goes live.
  - [x] This will only be done in servers that has the feature configured.
  - [x] The shout out message may be set by the server administrators.

### PR DESCRIPTION
**Pull request by MirageAegis**
# What Does This PR Do?
This PR brings configurable automatic go-live posts and shout outs to Sushi Bot.

## Acceptance Criteria Met
- Sushi Bot allows server administrators to configure auto shout outs.
- Sushi Bot shouts out users that are live on Twitch in servers with auto shout outs configured.
- Sushi Bot allows server administrators to configure auto go-live posts.
- Sushi Bot posts a go-live post when in servers where the server owner goes live on Twitch and has auto go-live posts configured.

## Related Issues
Closes #25

### CODE REVIEW

File: src/commands/configuration/config-subcommands/golive.ts [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+                       "Variables:\n" +
+                       `- \`${DISCORD_NAME}\` - Replaced with Discord display name\n` +
+                       `- \`${CHANNEL}\` - Replaced with Twitch username\n` +
+                       `- \`${LINK}\` - Replaced with Twitch link\n` +
+                       `- \`${TITLE}\` - Replaced with Stream title\n` +
+                       `- \`${GAME}\` - Replaced with Stream game\n` +
+                       `- \`${NEW_LINE}\` - Replaced with a new line`
+            }
+        )
+};
```
Comment (MirageAegis) [inline src/commands/configuration/config-subcommands/golive.ts:107]: This is the `/config golive` command. It lets server administrators configure a channel and message template to use when posting go-live posts in that channel.

File: src/commands/configuration/config-subcommands/shoutout.ts [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+                       "Variables:\n" +
+                       `- \`${DISCORD_NAME}\` - Replaced with Discord display name\n` +
+                       `- \`${CHANNEL}\` - Replaced with Twitch username\n` +
+                       `- \`${LINK}\` - Replaced with Twitch link\n` +
+                       `- \`${TITLE}\` - Replaced with Stream title\n` +
+                       `- \`${GAME}\` - Replaced with Stream game` +
+                       `- \`${NEW_LINE}\` - Replaced with a new line`
+            }
+        )
+};
```
Comment (MirageAegis) [inline src/commands/configuration/config-subcommands/shoutout.ts:115]: This is the `/config shoutout`. It lets server administrators configure auto shout outs by choosing a channel to post shout outs, a role to shout out, and a tessage template to use.

File: src/commands/moderation/ban.ts [hunk 3]
```diff
@@ -63,7 +63,7 @@ export const command: Command = {
 
         const member: GuildMember = <GuildMember> ctx.options.getMember("user") ?? null;
         const user: User = ctx.options.getUser("user"); // Never null
-        const reason: string = ctx.options.getString("reason") ?? null;
+        const reason: string = `${ctx.user}: ${ctx.options.getString("reason") ?? "No reason"}`;
```
Comment (MirageAegis) [inline src/commands/moderation/ban.ts:66]: This is added to all moderation commands to make them more transparent, showing the mod that used the commands.

File: src/events/events.ts [hunk 4]
```diff
@@ -34,40 +34,53 @@ import {
...
-        await logs.send({ embeds: [genMemberJoinEmbed(member)] });
-    } catch {
-        // Remove configurations if logging failed.
-        // Most likely causes are that the channel was deleted or that
-        // Sushi Bot had its permissions revoked
+    if (!logs) {
+        // Remove configurations if the logs do not exist
         server.logs = null;
         await server.save();
     }
```
Comment (MirageAegis) [inline src/events/events.ts:67]: The logic for logging and embed to a channel has been moved to this helper function.

File: src/events/events.ts [hunk 5]
```diff
@@ -310,31 +221,115 @@ export const onMessageEdit = async (client: Client, before: Message, after: Mess
...
-    // If embed is null, the message wasn't significant to Sushi Bot
-    if (!embed) {
+/**
+ * Broadcasts auto go-live posts and auto shout outs to sevrers subscribed to them.
+ * 
+ * @param client the Discord bot
+ * @param before the presence before update
+ * @param after the presence after update
+ */
+export const onPresenceUpdate = async (client: Client, before: Presence, after: Presence): Promise<void> => {
```
Comment (MirageAegis) [inline src/events/events.ts:241]: This function handles the few auto go-live and auto shout out features.

File: src/events/events.ts [hunk 6]
```diff
@@ -310,31 +221,115 @@ export const onMessageEdit = async (client: Client, before: Message, after: Mess
...
+    // Skip if no stream was found
+    if (!streams.length) {
+        return;
+    }
+    
+    // Check if the stream link has been updated
+    if (oldStreams.length) {
+        const oldUrls: string[] = oldStreams.map(o => o.url);
+        streams = streams.filter(s => !oldUrls.includes(s.url));
+    }
```
Comment (MirageAegis) [inline src/events/events.ts:286]: This part checks the difference of the activities before and after the presence change and proceeds if there's a unique Twitch stream after the presence change.

File: src/events/events.ts [hunk 7]
```diff
@@ -310,31 +221,115 @@ export const onMessageEdit = async (client: Client, before: Message, after: Mess
...
+            try {
+                await channel.send(formatGoLivePost(stream, server.goLive.message));
+            } catch (e) {
+                // Remove auto go-live configuration if we fail to send
+                server.goLive = null;
+                await server.save();
+                throw e;
+            }
+        }
+    } else {
```
Comment (MirageAegis) [inline src/events/events.ts:308]: If it's a server owner going live and their server has auto go-live posts configured, make a go-live post.

File: src/events/events.ts [hunk 8]
```diff
@@ -310,31 +221,115 @@ export const onMessageEdit = async (client: Client, before: Message, after: Mess
...
+            try {
+                await channel.send(formatGoLivePost(stream, server.shoutout.message));
+            } catch (e) {
+                // Remove auto shout out configuration if we fail to send
+                server.shoutout = null;
+                await server.save();
+                throw e;
+            }
+        }
     }
```
Comment (MirageAegis) [inline src/events/events.ts:333]: Otherwise, make a shout out post if auto shout outs are configured.

File: src/events/shoutout.ts [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+    const title: string = activity.details;
+    const game: string = activity.state;
+
+    return template.replace(DISCORD_NAME_RE, displayName)
+        .replace(CHANNEL_RE, channel)
+        .replace(LINK_RE, link)
+        .replace(TITLE_RE, title)
+        .replace(GAME_RE, game)
+        .replace(NEW_LINE_RE, "\n");
+};
```
Comment (MirageAegis) [inline src/events/shoutout.ts:113]: These functions format the message templates into go-live- or shout out posts.

File: src/events/shoutout.ts [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+ * The string that gets replaced with a new line character
+ * in the formatted string.
+ */
+export const NEW_LINE: string = "{nl}";
+const NEW_LINE_RE: RegExp = /{nl}/g;
+
+/**
+ * Default go-live and shout out stringused when none is in the configurations.
+ */
+const DEFAULT_POST: string = `${DISCORD_NAME} is live now! Go watch the stream at ${LINK}`;
```
Comment (MirageAegis) [inline src/events/shoutout.ts:76]: This is the template used if one isn't provided in the configurations.

File: src/schemas/server.ts [hunk 11]
```diff
@@ -27,8 +27,41 @@ import mongoose, { Schema, Document, InferSchemaType } from "mongoose";
...
+/**
+ * Auto go-live post configuration struct.
+ * 
+ * @param channel the channel ID of the auto go-live post channel
+ * @param message the auto go-live message
+ */
+export type GoLive = {
+    readonly channel?: string;
+    readonly message?: string;
+};
```
Comment (MirageAegis) [inline src/schemas/server.ts:63]: This is the new structure for server configurations with the added features.

File: src/events/logs.ts [hunk 12]
```diff
@@ -88,19 +127,30 @@ export const genMemberLeaveEmbed = (member: GuildMember): EmbedBuilder => {
...
-export const genMemberBanEmbed = (ban: GuildBan): EmbedBuilder => {
+export const genMemberBanEmbed = async (ban: GuildBan): Promise<EmbedBuilder> => {
+    // Get the ban data from the audit log
+    const logEntry: GuildAuditLogs<AuditLogEvent.MemberBanAdd> = await ban.guild.fetchAuditLogs({
+        limit: 1,
+        type: AuditLogEvent.MemberBanAdd,
+    });
+
+    // The most recent ban data
+    const banAction: GuildAuditLogsEntry = logEntry.entries.first();
```
Comment (MirageAegis) [inline src/events/logs.ts:138]: Additional enhancement: moderation actions have improved embeds.

**Review by MirageAegis — COMMENTED** [review]
This review goes through the key points of this PR, including the new auto shout out- and go-live post features.

**Review by zahatikoff — APPROVED** [review]

**Pull request merged**